In [136]:
%matplotlib inline
from pathlib import Path
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
import matplotlib.pylab as plt
from dmba import plotDecisionTree, classificationSummary, regressionSummary
import time 
import scipy.stats as stats

# Which major has the highest cumulative GPA?

In [137]:
# create dataframe from StudentClasses and set StudentID as index
student_class_df = pd.read_csv('StudentClass.csv')
student_class_df.set_index('StudentID', inplace=True)
student_class_df = student_class_df.dropna()

In [138]:
# create dataframe from StudentGraduate and set StudentID as index
student_grad_df = pd.read_csv('StudentGraduate.csv')
student_grad_df.set_index('ID', inplace=True)
student_grad_df = student_grad_df.dropna()

In [139]:
# replace grades with quantative values
# A = 4, B = 3, C = 2, D =1, F = 0
student_class_df.loc[student_class_df['GRADE'] == 'A', 'GRADE'] = '4'
student_class_df.loc[student_class_df['GRADE'] == 'B', 'GRADE'] = '3'
student_class_df.loc[student_class_df['GRADE'] == 'C', 'GRADE'] = '2'
student_class_df.loc[student_class_df['GRADE'] == 'D', 'GRADE'] = '1'
student_class_df.loc[student_class_df['GRADE'] == 'F', 'GRADE'] = '0'

In [140]:
# convert GRADE to integer as the former datatype will be string
student_class_df['GRADE'] = student_class_df['GRADE'].astype('int')

In [141]:
# create a new dataframe that shows each student, their grade, and the major
# other columns are irrelevant so drop them
students_df = pd.merge(student_class_df, student_grad_df['Major'], left_index=True, right_index=True)
students_df.drop(['ClassWeekDay', 'ClassMeetTime', 'ClassDate'], axis=1, inplace=True)

In [142]:
# find students who took math, physics, and statistics
quant_classes_df = students_df[
    (students_df.ClassTitle == 'Math') | (students_df.ClassTitle == 'Physics') | (students_df.ClassTitle == 'Statistics')]

In [144]:
# calculate gpas based on only math, physics, and statistics 
# group by gpa

quant_classes_df = quant_classes_df.groupby('Major')['GRADE'].mean()
quant_classes_df = quant_classes_df.to_frame()
quant_classes_df = quant_classes_df.rename(columns={'GRADE': 'GPA'}) 
quant_classes_df.sort_values('GPA', ascending=False)

GPA
Major          
MKTG   2.814246
MGMT   2.774970
I S    2.760000
FIN    2.749658
ACCT   2.726148

# Conclusion
Based on the students in various majors who took the quantitative classes of Math, Physics, and Statistics, we can conclude that marketing majors had the highest GPA of 2.81 based on the three quantitative classes only.